### PR TITLE
 sqlite3: Allow users to disable implicit transactions 

### LINKED ISSUE #346: sqlite: Migrations are always wrapped in implicit transactions
Contrary to the documentation and maintainer notes in the bug tracker -- see [here](https://github.com/golang-migrate/migrate/blob/777550e4d7cad06b865561956ca15750e8d8a2bf/GETTING_STARTED.md) and [here](https://github.com/golang-migrate/migrate/issues/196#issuecomment-494844896), for example -- the migrate database driver for SQLite does, in fact, wrap migration statements in an [implicit transaction](https://github.com/golang-migrate/migrate/blob/777550e4d7cad06b865561956ca15750e8d8a2bf/database/sqlite3/sqlite3.go#L186-L201).

These implicit transactions are _usually_ not a problem in SQLite.  Unlike other SQL implementations, DDL statements in SQLite can execute just fine within a transaction.  (Actually, in SQLite, a DDL statement that is executed outside a transaction will [automatically start a new transaction](https://www.sqlite.org/lang_transaction.html).  All DDL statements execute within a transaction.)  However, these implicit transactions do cause problems in other areas.

Users may be confused to find that their own `BEGIN` statements are met with errors.  A transaction may not be started within a transaction.  This is not a significant problem in and of itself.  The documentation could simply be amended to direct SQLite users to avoid their own `BEGIN` and `COMMIT` statements.

The big, and perhaps insurmountable, problem is that certain necessary SQLite `PRAGMA` statements cannot be executed within a transaction.  Due to limitations in SQLite's [`ALTER TABLE`](https://www.sqlite.org/lang_altertable.html) implementation, users often need to selectively toggle the [enforcement of foreign key constraints](https://www.sqlite.org/pragma.html#pragma_foreign_keys) when [pivoting between tables](https://www.sqlite.org/lang_altertable.html#making_other_kinds_of_table_schema_changes).  In wrapping all migrations with an implicit transaction, the migrate tool has snapped this essential ability away from the user.

One potential workaround would be for users to `COMMIT` the implicit transaction then `BEGIN` an explicit transaction within a migration.  I do not know whether this is safe.  To cite one source, the authors of [go-database-sql.org](https://web.archive.org/web/20200221175352/http://go-database-sql.org/modifying.html) have gone out of their way to caution users against mixing `Tx` methods with explicit `BEGIN`/`COMMIT` statements.

I am unsure whether current behaviour was intentional.  Unfortunately, I don't think it is possible to change this behaviour without breaking backwards compatibility.  Any users that are currently employing the workaround described in the previous paragraph would be broken by the removal of the implicit transactions.

Something for v5? :-)

---

**Migrate Version**
v4.7.1

**Loaded Source Drivers**
file

**Loaded Database Drivers**
sqlite3

**Go Version**
go version go1.13.8 darwin/amd64

### PR DESCRIPTION
**Pull request by saj**
Fixes #346.

### CODE REVIEW

File: database/sqlite3/sqlite3.go [hunk 1]
```diff
@@ -100,13 +102,22 @@ func (m *Sqlite) Open(url string) (database.Driver, error) {
...
+	qv := purl.Query()
+
+	migrationsTable := qv.Get("x-migrations-table")
 	if len(migrationsTable) == 0 {
 		migrationsTable = DefaultMigrationsTable
 	}
+
+	noTxWrap, err := strconv.ParseBool(qv.Get("x-no-tx-wrap"))
+	if err != nil {
+		noTxWrap = false
```
Comment (dhui) [inline database/sqlite3/sqlite3.go]: It's cleaner to return an error for invalid values. We should handle non-existent (e.g. unspecified) values separately.

Comment (saj) [reply database/sqlite3/sqlite3.go]: I did wonder about this, so I went looking for existing patterns.  I found this behaviour in the [cockroachdb driver](https://github.com/golang-migrate/migrate/blob/6dc95bd2010d471d36c59629bce101fc916677e5/database/cockroachdb/cockroachdb.go#L130-L133).  Would you still prefer an error -- at the cost of consistency -- in this case?

Comment (dhui) [reply database/sqlite3/sqlite3.go]: Yeah, it's cleaner to fail for invalid inputs instead of silently assuming a value

Comment (saj) [reply database/sqlite3/sqlite3.go]: I have pushed a commit with the change you requested.  PTAL, @dhui.

**Pull request merged**